add new Odesli/Songlink domains

kweiberth opened the pull request:
#### Changes proposed in this Pull Request:
- added new Songlink/Odesli domains. Songlink used to run only at https://song.link, but now it supports a handleful of other pretty domains, such as album.link and artist.link. As part of these changes, Songlink rebranded to Odesli, with song.link, album.link, etc. now different domains under Odesli name. 
- https://song.link/oembed is still supported (and always will be), but figured it's better to use the top-level/main domain https://odesli.co for oEmebd endpoint
- Let me know if you'd like me to provide any example URLs for testing; e.g. https://song.link/i/1051394215, https://album.link/i/1051394208, https://artist.link/mars. I can give you a fuller list of all possibilities/types of URLs
- Original whitelisting of Songlink was tracked [here](https://github.com/Automattic/jetpack/issues/9973). 

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
- Enhances support of a pre-existing feature

#### Testing instructions:
- The [original PR](https://github.com/Automattic/jetpack/pull/12954/files) that added whitelist support did not add tests, but if you show me how to add tests for this, I can do so.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
- Added support for all domains supported by Odesli (formerly Songlink): song.link, album.link, artist.link, playlist.link, pods.link, mylink.page, and odesli.co

kweiberth commented:
Looks like there's some config/auth issue with e2e test step:

```
$ openssl enc -md sha1 -aes-256-cbc -d -pass env:CONFIG_KEY -in ./tests/e2e/config/encrypted.enc -out ./tests/e2e/config/local-test.js
Can't read environment variable CONFIG_KEY
Error getting password
error Command failed with exit code 1.
```

jeherve reviewed (changes requested):
You seem to have a mix or spaces and tabs in there. I'd be best to switch to tabs only. I think we can also remove the comment here, as one can get to the Pull Request by `git blame`ing the file.

In addition to this, I would recommend removing "staging", as it's not really something that we need to support or look for in links on production sites.

> Let me know if you'd like me to provide any example URLs for testing

That would be great, thank you!

kweiberth commented:
All sounds good. I've removed the comment, used tabs only, and removed the support for staging. I added staging bc for us staging is production-like and, should we ever want to debug our embed in a Wordpress site, would be easier to try and test changes to our staging env than our production. But all should be good, I don't expect any issues. 

@jeherve How many URLs would you like above and beyond the few I provided in the PR description? With different types of URLs and different domains, the permutation adds up to quite a few different URLs.

jeherve commented:
> With different types of URLs and different domains, the permutation adds up to quite a few different URLs.

Do you think you could build a list with one of each, just like something we could use in a unit test to check all permutations at once?

kweiberth commented:
Definitely, he's a complete list: 

```
https://song.link/hello 
https://album.link/hello
https://pods.link/hello
https://odesli.com/hello
https://odesli.co/hello
https://artist.link/hello
https://playlist.link/hello
https://mylink.page/hello
https://song.link/ca/i/1288547996
https://song.link/song/ca/i/1288547996
https://album.link/song/ca/i/1288547996
https://odesli.com/song/ca/i/1288547996
https://odesli.co/song/ca/i/1288547996
https://artist.link/song/ca/i/1288547996
https://playlist.link/song/ca/i/1288547996
https://mylink.page/song/ca/i/1288547996
https://song.link/i/1288547996
https://song.link/song/i/1288547996
https://album.link/song/i/1288547996
https://odesli.com/song/i/1288547996
https://odesli.co/song/i/1288547996
https://artist.link/song/i/1288547996
https://playlist.link/song/i/1288547996
https://mylink.page/song/i/1288547996
https://song.link/song/s/6r1siiMXYLrao42oiqQrgj
https://album.link/song/s/6r1siiMXYLrao42oiqQrgj
https://odesli.com/song/s/6r1siiMXYLrao42oiqQrgj
https://odesli.co/song/s/6r1siiMXYLrao42oiqQrgj
https://artist.link/song/s/6r1siiMXYLrao42oiqQrgj
https://playlist.link/song/s/6r1siiMXYLrao42oiqQrgj
https://mylink.page/song/s/6r1siiMXYLrao42oiqQrgj
https://song.link/s/6r1siiMXYLrao42oiqQrgj
https://song.link/song/y/UUDjpNpETzo
https://album.link/song/y/UUDjpNpETzo
https://odesli.com/song/y/UUDjpNpETzo
https://odesli.co/song/y/UUDjpNpETzo
https://artist.link/song/y/UUDjpNpETzo
https://playlist.link/song/y/UUDjpNpETzo
https://mylink.page/song/y/UUDjpNpETzo
https://song.link/y/UUDjpNpETzo
https://song.link/song/y/6366dxFf-Os
https://album.link/song/y/6366dxFf-Os
https://odesli.com/song/y/6366dxFf-Os
https://odesli.co/song/y/6366dxFf-Os
https://artist.link/song/y/6366dxFf-Os
https://playlist.link/song/y/6366dxFf-Os
https://mylink.page/song/y/6366dxFf-Os
https://song.link/y/6366dxFf-Os
https://song.link/song/y/H8tLS_NOWLs
https://album.link/song/y/H8tLS_NOWLs
https://odesli.com/song/y/H8tLS_NOWLs
https://odesli.co/song/y/H8tLS_NOWLs
https://artist.link/song/y/H8tLS_NOWLs
https://playlist.link/song/y/H8tLS_NOWLs
https://mylink.page/song/y/H8tLS_NOWLs
https://song.link/y/H8tLS_NOWLs
https://song.link/song/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://album.link/song/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://odesli.com/song/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://odesli.co/song/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://artist.link/song/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://playlist.link/song/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://mylink.page/song/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://song.link/g/Tg6zuj3cjgwcgrdi6ewlqdjk6iq
https://song.link/song/d/138241229
https://album.link/song/d/138241229
https://odesli.com/song/d/138241229
https://odesli.co/song/d/138241229
https://artist.link/song/d/138241229
https://playlist.link/song/d/138241229
https://mylink.page/song/d/138241229
https://song.link/d/138241229
https://song.link/song/t/72268642
https://album.link/song/t/72268642
https://odesli.com/song/t/72268642
https://odesli.co/song/t/72268642
https://artist.link/song/t/72268642
https://playlist.link/song/t/72268642
https://mylink.page/song/t/72268642
https://song.link/t/72268642
https://song.link/song/a/B01NAE38YO
https://album.link/song/a/B01NAE38YO
https://odesli.com/song/a/B01NAE38YO
https://odesli.co/song/a/B01NAE38YO
https://artist.link/song/a/B01NAE38YO
https://playlist.link/song/a/B01NAE38YO
https://mylink.page/song/a/B01NAE38YO
https://song.link/a/B01NAE38YO
https://song.link/song/n/tra.246588040
https://album.link/song/n/tra.246588040
https://odesli.com/song/n/tra.246588040
https://odesli.co/song/n/tra.246588040
https://artist.link/song/n/tra.246588040
https://playlist.link/song/n/tra.246588040
https://mylink.page/song/n/tra.246588040
https://song.link/n/tra.246588040
https://song.link/song/sc/347741803
https://album.link/song/sc/347741803
https://odesli.com/song/sc/347741803
https://odesli.co/song/sc/347741803
https://artist.link/song/sc/347741803
https://playlist.link/song/sc/347741803
https://mylink.page/song/sc/347741803
https://song.link/sc/347741803
https://song.link/song/ya/32504596
https://album.link/song/ya/32504596
https://odesli.com/song/ya/32504596
https://odesli.co/song/ya/32504596
https://artist.link/song/ya/32504596
https://playlist.link/song/ya/32504596
https://mylink.page/song/ya/32504596
https://song.link/ya/32504596
https://song.link/song/p/TR:5791220
https://album.link/song/p/TR:5791220
https://odesli.com/song/p/TR:5791220
https://odesli.co/song/p/TR:5791220
https://artist.link/song/p/TR:5791220
https://playlist.link/song/p/TR:5791220
https://mylink.page/song/p/TR:5791220
https://song.link/p/TR:5791220
https://song.link/song/sp/1767083
https://album.link/song/sp/1767083
https://odesli.com/song/sp/1767083
https://odesli.co/song/sp/1767083
https://artist.link/song/sp/1767083
https://playlist.link/song/sp/1767083
https://mylink.page/song/sp/1767083
https://song.link/sp/1767083
https://album.link/ca/i/1279947491
https://song.link/album/ca/i/1279947491
https://album.link/album/ca/i/1279947491
https://odesli.com/album/ca/i/1279947491
https://odesli.co/album/ca/i/1279947491
https://artist.link/album/ca/i/1279947491
https://playlist.link/album/ca/i/1279947491
https://mylink.page/album/ca/i/1279947491
https://album.link/i/1279947491
https://song.link/album/i/1279947491
https://album.link/album/i/1279947491
https://odesli.com/album/i/1279947491
https://odesli.co/album/i/1279947491
https://artist.link/album/i/1279947491
https://playlist.link/album/i/1279947491
https://mylink.page/album/i/1279947491
https://album.link/s/6JckISQKS5IwOH4YrFrfIB
https://song.link/album/s/6JckISQKS5IwOH4YrFrfIB
https://album.link/album/s/6JckISQKS5IwOH4YrFrfIB
https://odesli.com/album/s/6JckISQKS5IwOH4YrFrfIB
https://odesli.co/album/s/6JckISQKS5IwOH4YrFrfIB
https://artist.link/album/s/6JckISQKS5IwOH4YrFrfIB
https://playlist.link/album/s/6JckISQKS5IwOH4YrFrfIB
https://mylink.page/album/s/6JckISQKS5IwOH4YrFrfIB
https://album.link/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://song.link/album/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://album.link/album/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://odesli.com/album/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://odesli.co/album/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://artist.link/album/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://playlist.link/album/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://mylink.page/album/y/OLAK5uy_nErvzigwPVzj76NSzpte5pmiViP_F0gNE
https://album.link/g/B7xsarspqwsinxzont6zdvd2hny
https://song.link/album/g/B7xsarspqwsinxzont6zdvd2hny
https://album.link/album/g/B7xsarspqwsinxzont6zdvd2hny
https://odesli.com/album/g/B7xsarspqwsinxzont6zdvd2hny
https://odesli.co/album/g/B7xsarspqwsinxzont6zdvd2hny
https://artist.link/album/g/B7xsarspqwsinxzont6zdvd2hny
https://playlist.link/album/g/B7xsarspqwsinxzont6zdvd2hny
https://mylink.page/album/g/B7xsarspqwsinxzont6zdvd2hny
https://album.link/d/14843037
https://song.link/album/d/14843037
https://album.link/album/d/14843037
https://odesli.com/album/d/14843037
https://odesli.co/album/d/14843037
https://artist.link/album/d/14843037
https://playlist.link/album/d/14843037
https://mylink.page/album/d/14843037
https://album.link/t/67784530
https://song.link/album/t/67784530
https://album.link/album/t/67784530
https://odesli.com/album/t/67784530
https://odesli.co/album/t/67784530
https://artist.link/album/t/67784530
https://playlist.link/album/t/67784530
https://mylink.page/album/t/67784530
https://album.link/a/B01N48U32A
https://song.link/album/a/B01N48U32A
https://album.link/album/a/B01N48U32A
https://odesli.com/album/a/B01N48U32A
https://odesli.co/album/a/B01N48U32A
https://artist.link/album/a/B01N48U32A
https://playlist.link/album/a/B01N48U32A
https://mylink.page/album/a/B01N48U32A
https://album.link/n/alb.246588025
https://song.link/album/n/alb.246588025
https://album.link/album/n/alb.246588025
https://odesli.com/album/n/alb.246588025
https://odesli.co/album/n/alb.246588025
https://artist.link/album/n/alb.246588025
https://playlist.link/album/n/alb.246588025
https://mylink.page/album/n/alb.246588025
https://album.link/sc/228667052
https://song.link/album/sc/228667052
https://album.link/album/sc/228667052
https://odesli.com/album/sc/228667052
https://odesli.co/album/sc/228667052
https://artist.link/album/sc/228667052
https://playlist.link/album/sc/228667052
https://mylink.page/album/sc/228667052
https://album.link/ya/297720
https://song.link/album/ya/297720
https://album.link/album/ya/297720
https://odesli.com/album/ya/297720
https://odesli.co/album/ya/297720
https://artist.link/album/ya/297720
https://playlist.link/album/ya/297720
https://mylink.page/album/ya/297720
https://album.link/p/AL:668238
https://song.link/album/p/AL:668238
https://album.link/album/p/AL:668238
https://odesli.com/album/p/AL:668238
https://odesli.co/album/p/AL:668238
https://artist.link/album/p/AL:668238
https://playlist.link/album/p/AL:668238
https://mylink.page/album/p/AL:668238
https://album.link/sp/145895
https://song.link/album/sp/145895
https://album.link/album/sp/145895
https://odesli.com/album/sp/145895
https://odesli.co/album/sp/145895
https://artist.link/album/sp/145895
https://playlist.link/album/sp/145895
https://mylink.page/album/sp/145895
https://pods.link/us/i/360084272
https://song.link/podcast/us/i/360084272
https://album.link/podcast/us/i/360084272
https://odesli.com/podcast/us/i/360084272
https://odesli.co/podcast/us/i/360084272
https://artist.link/podcast/us/i/360084272
https://playlist.link/podcast/us/i/360084272
https://mylink.page/podcast/us/i/360084272
https://pods.link/i/360084272
https://song.link/podcast/i/360084272
https://album.link/podcast/i/360084272
https://odesli.com/podcast/i/360084272
https://odesli.co/podcast/i/360084272
https://artist.link/podcast/i/360084272
https://playlist.link/podcast/i/360084272
https://mylink.page/podcast/i/360084272
```

Most of these should be valid links, though some might not exist and would 404 if you tried to load them. However, the structures and path segments are indicative of our supported links.

jeherve commented:
I rebased to fix a problematic Unit Test. Let's give this another review now. 👍

kweiberth commented:
Awesome. I believe I've made all requested changes. Let me know if there's anything else you'd like me to do.

leogermani reviewed (changes requested):
Hi @kweiberth ,

Thanks for the PR.

I've tested it and it didn't work for me. Have you tested it?

Digging it a bit I found out that, since the pattern you are registering is in a regex format, you must inform it to `wp_oembed_add_provider` function, setting the third parameter to `true`. 

I got this working with the following line:

```PHP
wp_oembed_add_provider( '#https?://((song|album|artist|pods|playlist)\.link|odesli\.com?|mylink\.page)/.*#', 'https://odesli.co/oembed', true );
```
It would be awesome if you could write some tests to assure that your regex is working with all expected URLs. For this, your test could be done against `WP_oEmbed::get_provider()`:

```PHP
// ...
$oembed = _wp_oembed_get_object();
$test_url = 'https://song.link/y/UUDjpNpETzo';
$args = array(
	'discover' => false // do not make any request
);
$expected = 'https://odesli.co/oembed';

$this->assertEquals( $expected, $oembed->get_provider( $test_url, $args ) );
```
You could make a loop and test all that URL list you've posted as a comment! Feel free to create a new file for this test inside the `tests` folder as I don't think there are any other similar test yet.

**Note** The "Test Instructions" section in our PR template refers to instructions to help human reviewers test the PR. So, for example, in this case it would be something like: "Create a post and insert one of the following links to the content. Check that the link is converted into an embed"

Cheers

kweiberth commented:
@leogermani thank you for looking into this and for providing info re: tests. Will happily write tests. Will let you know if I have any questions as I figure out PHP :)

kweiberth commented:
@leogermani @jeherve I need some help running PHP unit tests. I manually install phpunit, and I'm trying to run:

```sh
./phpunit -c tests/php.multisite.xml
```

But I'm getting this error: 

```sh
Using test root
To run Jetpack woocommerce tests, prefix phpunit with JETPACK_TEST_WOOCOMMERCE=1

Warning: require(/includes/functions.php): failed to open stream: No such file or directory in /Users/Kurt/code/jetpack/tests/php/bootstrap.php on line 66

Fatal error: require(): Failed opening required '/includes/functions.php' (include_path='.:') in /Users/Kurt/code/jetpack/tests/php/bootstrap.php on line 66
```

Any suggestions for next steps? My main goal right now is to run a specific test file so that I can then write/debug the test(s) I want to add.

jeherve commented:
@kweiberth In #15099, I am moving the function that adds those embeds into a different location in the codebase. I also added an example test. All this should  make it easier for you to build your own test in the future, once that PR is merged.

To run the tests locally, you have 2 options:
- If you use our Docker instance as your development environment, it should be fairly straightforward:
https://github.com/Automattic/jetpack/blob/ab5cd45ad5f9f1488c9899134386902609a61ae1/docker/README.md#L139
- If you don't, things are a bit more involved: 
https://github.com/Automattic/jetpack/blob/648c82fc10701a084e9dae0cd273588d77db0248/docs/development-environment.md#L89

kweiberth commented:
@jeherve that PR and the extra info is incredibly helpful, thank you. I will work on rebasing this and adding a test that runs for all our URL structures

kweiberth commented:
Hello @jeherve, apologies for the delay on this, but I'm picking this back up so hopefully we can get these new domains supported. 

I was able to `yarn docker:up` but I'm struggling to run the unit tests. When I `yarn docker:phpunit` I get the following error output:

```
yarn docker:phpunit

yarn run v1.15.2
$ yarn docker:compose exec wordpress phpunit --configuration=/var/www/html/wp-content/plugins/jetpack/phpunit.xml.dist
$ yarn docker:env && yarn docker:compose-volumes && yarn docker:compose-extras && docker-compose -f docker/docker-compose.yml -f docker/compose-volumes.built.yml -f docker/compose-extras.yml exec wordpress phpunit --configuration=/var/www/html/wp-content/plugins/jetpack/phpunit.xml.dist
$ node -e "require('fs').createWriteStream( 'docker/.env', { flags: 'a' } );"
$ [ -f docker/compose-volumes.yml ] || cp docker/compose-volumes.yml.sample docker/compose-volumes.yml; bash ./bin/compose-volumes.sh docker/compose-volumes.yml > docker/compose-volumes.built.yml
$ [ -f docker/compose-extras.yml ] || cp docker/compose-extras.yml.sample docker/compose-extras.yml
Using test root /tmp/wordpress-develop/tests/phpunit
To run Jetpack woocommerce tests, prefix phpunit with JETPACK_TEST_WOOCOMMERCE=1
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
error Command failed with exit code 255.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 255.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Any ideas on what's going on? There doesn't seem to be a specific/useful error message

kweiberth committed:
add comment so I can reopen PR

jeherve commented:
> yarn docker:phpunit

I would recommend running only the tests you're changing, like so:

`yarn docker:phpunit --filter= WP_Test_Jetpack_Shortcodes_Others`

If you still run into issues, it may be worth trying to update the core unit tests suite on your environment:

`yarn docker:update-core-unit-tests`

You should then be able to start your container again and then run the tests.

Let me know how that goes.

kweiberth commented:
Hey @jeherve I was able to run

`yarn docker:update-core-unit-tests` 

and then 

`yarn docker:phpunit --filter=WP_Test_Jetpack_Shortcodes_Others`

and was able to get the test to run and pass. However, I then a code change to the test to try and add an array of URLs, and then I was unable to run the tests. No combination of updating, stopping/restarting containers, etc. helped me. I kept getting the same error as I posted above:

```
error Command failed with exit code 255.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

If I provided you a PHP array of all the URLs we support, like I provided in [the past](https://github.com/Automattic/jetpack/pull/14998#issuecomment-600367587), would you be able to modify this test to loop through the array of URLs? Unfortunately I'm not fluent in PHP and I'm having a hard time getting the environment up and running. I have a feeling it would take you 5 mins and me 5 hours (at least). 

If so, is this how I would define an array of URLs?

```php
$array(
    'https://song.link/i/123',
    'https://song.link/i/abc'
);
```

jeherve commented:
I created #17095 to update the embed provider and add some tests. Let me know what you think.
